File: problems/cpp/convert-sorted-array-to-binary-search-tree.cc
// 12 ms, 09/09/2017
// Time  : O(n)
// Space : O(h)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sorted_range_to_bst(const vector<int>& nums, int first, int last)
    {
        if (first > last)
            return nullptr;
        int mid = first + (last - first) / 2;
        auto root = new TreeNode(nums[mid]);
        root->left = sorted_range_to_bst(nums, first, mid - 1);
        root->right = sorted_range_to_bst(nums, mid + 1, last);
        return root;
    }
    TreeNode* sortedArrayToBST(vector<int>& nums)
    {
        return sorted_range_to_bst(nums, 0, nums.size() - 1);
    }
};

File: code/python/convert-sorted-array-to-binary-search-tree.py
# Date  : 03/25/2018
# Time  : O(n)
# Space : O(lg(n))

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        def build_bst(index_left, index_past_right):
            if index_left == index_past_right:
                return None
            index_root = (index_left + index_past_right) // 2
            root = TreeNode(nums[index_root])
            root.left = build_bst(index_left, index_root)
            root.right = build_bst(index_root + 1, index_past_right)
            return root

        return build_bst(0, len(nums))
